azurelinuxagent/common/cgroupstelemetry.py

```diff
@@ -13,47 +13,184 @@
 # limitations under the License.
 #
 # Requires Python 2.6+ and Openssl 1.0+
+import threading
+from datetime import datetime as dt
 
-import time
-
-from azurelinuxagent.common import logger
-from azurelinuxagent.common.cgroup import CpuCgroup, MemoryCGroup
 
 class CGroupsTelemetry(object):
     """
     """
-    _tracked = {}
+    _tracked = []
+    _cgroup_metrics = {}
+    _rlock = threading.RLock()
+
+    @staticmethod
+    def _get_metrics_list(metric):
+        return [metric.average(), metric.min(), metric.max(), metric.median(), metric.count(), metric.first_polltime(),
+                metric.last_polltime()]
+
+    @staticmethod
+    def _process_cgroup_metric(cgroup_metrics):
+        current_memory_usage, max_memory_levels, current_cpu_usage = cgroup_metrics.get_metrics()
+        processed_extension = {
+            "memory": {
+                "cur_mem": CGroupsTelemetry._get_metrics_list(current_memory_usage),
+                "max_mem": CGroupsTelemetry._get_metrics_list(max_memory_levels)
+            },
+            "cpu": {
+                "cur_cpu": CGroupsTelemetry._get_metrics_list(current_cpu_usage)
+            }
+        }
+        return processed_extension
 
     @staticmethod
     def metrics_hierarchies():
-        return CGroupsTelemetry._hierarchies
+        pass  # return CGroupsTelemetry._hierarchies
 
     @staticmethod
     def track_cgroup(cgroup):
         """
-        Adds the given item to the list of tracked cgroups
+        Adds the given item to the dictionary of tracked cgroups
         """
-        pass  # TODO
+        # Making track thread-safe.
+        CGroupsTelemetry._rlock.acquire()
```

Comment:
let's acquire/release the lock using try/finally

Reply:
Changed this to use `with` as suggested in [here](https://stackoverflow.com/a/25704099). This takes care of releasing the lock if we hit an issue.

Thanks for catching this.